httpbp/response.go

```diff
@@ -0,0 +1,210 @@
+	if e.ResponseMessage != "" {
+		return e.ResponseMessage
+	}
+	return http.StatusText(e.StatusCode())
+}
+
+// As implements helper interface for errors.As.
+//
+// If v is pointer to either HTTPError or *HTTPError,
+// *v will be set into this error.
+// Otherwise, As will try errors.As against the Cause of the HTTPError
+func (e HTTPError) As(v interface{}) bool {
+	if target, ok := v.(*HTTPError); ok {
+		*target = e
+		return true
+	} else if target, ok := v.(**HTTPError); ok {
+		*target = &e
+		return true
+	}
+	return errors.As(e.Cause, v)
+}
+
+// Is implements helper interface for errors.Is.
+//
+// Returns true if target is equal to the HTTPError.
+// Otherwise, Is will try errors.Is against the Cause of the HTTPError
+func (e HTTPError) Is(target error) bool {
+	if err, ok := target.(HTTPError); ok {
+		return err == e
+	} else if err, ok := target.(*HTTPError); ok {
+		return *err == e
+	}
+	return errors.Is(e.Cause, target)
+}
+
+// Unwrap implements helper interface for errors.Unwrap.  Returns the optional
+// e.Cause error.
+func (e HTTPError) Unwrap() error {
+	return e.Cause
+}
+
+// StatusCode returns the HTTP status code to set on the response.  Defaults to
+// 500 if Code is not set on the HTTPError.
+func (e HTTPError) StatusCode() int {
+	if e.Code == 0 {
+		return http.StatusInternalServerError
+	}
+	return e.Code
+}
+
+var (
+	_ error = HTTPError{}
+	_ error = (*HTTPError)(nil)
+)
+
+// ResponseCookies is an interface that your Response objects can implement in
+// order to have the httpbp.Encode methods automatically add cookies to the
+// response.
+type ResponseCookies interface {
+	// Return a list of all cookies to set on the response.
+	Cookies() []*http.Cookie
+}
+
+// ResponseHeaders is an interface that your Response objects can implement in
+// order to have the httpbp.Encode methods automatically add http headers to the
+// response.
+type ResponseHeaders interface {
+	// Headers returns the http.Header container that can be used to add headers
+	// to the response.
+	Headers() http.Header
+}
+
+// ResponseCode is an interface that your Response objects can implement in
+// order to have the httpbp.Encode methods automatically set the response code
+// to something other than 200.
+//
+// If you implement the ErrorResponse interface, any error response codes will
+// be based on the ErrorMapper rather than the StatusCode method implemented in
+// this interface.
+type ResponseCode interface {
+	// StatusCode returns the HTTP status code to set on the response.
+	StatusCode() int
+}
+
+// ErrorResponse is an interface that your Response objects can implement in
+// order to have the httpbp.Encode methods automatically return http errors.
+type ErrorResponse interface {
+	// Err returns the HTTPError set on the response.
+	//
+	// This method is named "Err" rather than "Error" to avoid confusion with
+	// the "Error" function that is specified in the error interface.
+	Err() *HTTPError
```

Comment:
I would prefer to not force the error type here. When using it we can just do a casting:

```go
if respErr := resp.Err(); respErr != nil {
  var he HTTPError
  if !errors.As(respErr, &he) {
    he.Code = 500
    he.Cause = respErr
  }
  http.Error(w, he.ResponseMessage(), he.StatusCode())
  return nil
}
```